### PR TITLE
Mat 6264 export used data ele def qdm

### PR DESCRIPTION
**Pull request by sb-cecilialiu**
## CQL to ELM Translation Service PR

Jira Ticket: [MAT-6264](https://jira.cms.gov/browse/MAT-6264)
(Optional) Related Tickets:

### All Submissions
* [x] This PR has the JIRA linked.
* [x] Required tests are included.
* [x] No extemporaneous files are included (i.e Complied files or testing results).
* [x] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [x] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/gov/cms/madie/qdm/humanreadable/model/HumanReadableCodeModel.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    this.codesystemName = taxonomy;
+  }
+
+  public boolean getIsCodesystemVersionIncluded() {
+    return isCodesystemVersionIncluded;
+  }
+
+  public void setIsCodesystemVersionIncluded(boolean isCodeSystemIncluded) {
+    this.isCodesystemVersionIncluded = isCodeSystemIncluded;
+  }
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/qdm/humanreadable/model/HumanReadableCodeModel.java]: Boilerplate setters can be removed since they'll be added by lombok's `@Data` annotation.

File: src/main/java/gov/cms/madie/qdm/humanreadable/model/HumanReadableCodeModel.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    this.codesystemVersion = codesystemVersion;
+  }
+
+  public String getDatatype() {
+    return datatype;
+  }
+
+  public void setDatatype(String datatype) {
+    this.datatype = datatype;
+  }
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/qdm/humanreadable/model/HumanReadableCodeModel.java]: Will be generated by lombok, can be removed.

File: src/main/java/gov/cms/mat/cql_elm_translation/service/DataCriteriaService.java [hunk 3]
```diff
@@ -213,4 +213,19 @@ private String buildCriteriaType(String dataType) {
...
+    CQLTools tools = parseCql(cql, accessToken, cqlLibraryService);
+    return tools.getUsedValuesets();
+  }
+
+  public List<CQLValueSet> getUsedCQLValuesets(String cql, String accessToken) {
+    CQLTools tools = parseCql(cql, accessToken, cqlLibraryService);
+    return tools.getUsedCQLValuesets();
+  }
+
+  public List<CQLCode> getCQLCodes(String cql, String accessToken) {
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/mat/cql_elm_translation/service/DataCriteriaService.java]: Might as well keep with the preceding naming pattern:
```suggestion
  public List<CQLCode> getUsedCQLCodes(String cql, String accessToken) {
```

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java [hunk 4]
```diff
@@ -98,57 +136,301 @@ HumanReadableMeasureInformationModel buildMeasureInfo(Measure measure) {
...
         .map(
             population ->
                 HumanReadablePopulationModel.builder()
                     .name(population.getName().name())
                     .id(population.getId())
                     .display(population.getName().getDisplay())
-                    .logic(population.getDefinition())
+                    .logic(getCQLDefinitionLogic(population.getDefinition(), allDefinitions))
                     .expressionName(population.getDefinition())
+                    .inGroup(StringUtils.isBlank(population.getDefinition()) ? false : true)
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: `isBlank` will return true/false without the ternary operator.
```suggestion
                    .inGroup(!StringUtils.isBlank(population.getDefinition()))
```

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java [hunk 5]
```diff
@@ -98,57 +136,301 @@ HumanReadableMeasureInformationModel buildMeasureInfo(Measure measure) {
...
+    return group.getStratifications().stream()
+        .map(
+            stratification ->
+                HumanReadablePopulationModel.builder()
+                    .name("Stratification")
+                    .id(stratification.getId())
+                    .display("Stratification")
+                    .logic(getCQLDefinitionLogic(stratification.getCqlDefinition(), allDefinitions))
+                    .expressionName(stratification.getCqlDefinition())
+                    .inGroup(StringUtils.isBlank(stratification.getCqlDefinition()) ? false : true)
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: Ternary can be removed.
```suggestion
                    .inGroup(!StringUtils.isBlank(stratification.getCqlDefinition()))
```

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java [hunk 6]
```diff
@@ -98,57 +136,301 @@ HumanReadableMeasureInformationModel buildMeasureInfo(Measure measure) {
...
+                            definition
+                                .getLogic()
+                                .substring(definition.getLogic().indexOf('\n') + 1))
+                        .build())
+            .collect(Collectors.toList());
+    expressions.sort(Comparator.comparing(HumanReadableExpressionModel::getName));
+    return expressions;
+  }
+
+  boolean findUsedFunction(Measure measure, String accessToken, String id) {
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: Recommend renaming with `is` prefix. `Find` methods are commonly expected to return an Object and 'is' methods return boolean.
```suggestion
  boolean isUsedFunction(Measure measure, String accessToken, String id) {
```

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java [hunk 7]
```diff
@@ -98,57 +136,301 @@ HumanReadableMeasureInformationModel buildMeasureInfo(Measure measure) {
...
+          cqlCodes.stream()
+              .map(
+                  cqlCode ->
+                      HumanReadableCodeModel.builder()
+                          .name(cqlCode.getCodeName())
+                          .oid(cqlCode.getId())
+                          .codesystemName(cqlCode.getCodeSystemName())
+                          .codesystemVersion(cqlCode.getCodeSystemVersion())
+                          .isCodesystemVersionIncluded(cqlCode.isIsCodeSystemVersionIncluded())
+                          .datatype(cqlCode.getCodeIdentifier()) // ???
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: Is there a question around datatype we should answer?

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java [hunk 8]
```diff
@@ -98,57 +136,301 @@ HumanReadableMeasureInformationModel buildMeasureInfo(Measure measure) {
...
+    List<CQLValueSet> usedValuesets =
+        dataCriteriaService.getUsedCQLValuesets(measure.getCql(), accessToken);
+    List<CQLValueSet> otherValueSets = new ArrayList<>();
+    if (!CollectionUtils.isEmpty(usedValuesets)) {
+      for (CQLValueSet cqlValueSet : usedValuesets) {
+        if (findNotExistInSourceDataCriteria(cqlValueSet.getOid(), sourceDataCriteria) == null) {
+          otherValueSets.add(cqlValueSet);
+        }
+      }
+    }
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: I tried to match the intent using stream filter. If it looks good, then `findNotExistInSourceDataCriteria` can be removed.
```suggestion
    List<CQLValueSet> otherValueSets =
        usedValuesets.stream()
            .filter(
                vs ->
                    sourceDataCriteria.stream()
                        .noneMatch(sdc -> sdc.getOid().equalsIgnoreCase(vs.getOid())))
            .toList();
```

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/cql/CQLTools.java [hunk 9]
```diff
@@ -226,10 +231,10 @@ private void collectUsedFunctions(
       CQLGraph graph, List<String> functions, String parentExpression) {
     for (String function : functions) {
       if (graph.isPath(parentExpression, function)) {
-        if (usedDefinitions.containsKey(function)) {
-          usedDefinitions.get(function).add(parentExpression);
+        if (usedFunctions.containsKey(function)) {
+          usedFunctions.get(function).add(parentExpression);
         } else {
-          usedDefinitions.putIfAbsent(
+          usedFunctions.putIfAbsent(
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/mat/cql_elm_translation/utils/cql/CQLTools.java:237]: I need to review this change more closely as it _may_ impact the determination of a CQL Define's Call Hierarchy. 

I think the call hierarchy processor considers CQL Functions to be CQL Defines since they are written as `define function ...`.

**Review by jkotanchik-SB — CHANGES_REQUESTED** [review]

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java (hunk 7 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: @jkotanchik-SB Yeah I'm not sure what to use to populate the datatype field?

File: src/main/java/gov/cms/madie/qdm/humanreadable/model/HumanReadableCodeModel.java (hunk 1 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/madie/qdm/humanreadable/model/HumanReadableCodeModel.java]: Done

File: src/main/java/gov/cms/madie/qdm/humanreadable/model/HumanReadableCodeModel.java (hunk 2 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/madie/qdm/humanreadable/model/HumanReadableCodeModel.java]: Done.

File: src/main/java/gov/cms/mat/cql_elm_translation/service/DataCriteriaService.java (hunk 3 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/DataCriteriaService.java]: Changed

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java (hunk 4 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: Changed

Comment (sb-cecilialiu) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: Changed

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java (hunk 5 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: Changed

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java (hunk 6 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: Changed

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java (hunk 8 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: Changed

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/cql/CQLTools.java (hunk 9 above)
Comment (sb-cecilialiu) [reply src/main/java/gov/cms/mat/cql_elm_translation/utils/cql/CQLTools.java:237]: @jkotanchik-SB Ok thanks. This is the one used to populate Functions section.

File: src/main/java/gov/cms/madie/qdm/humanreadable/model/HumanReadableCodeModel.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+  @Override
+  public String getDataCriteriaDisplay() {
+    createDataCriteriaDisplay();
+    return dataCriteriaDisplay;
+  }
+
+  @Override
+  public void setDataCriteriaDisplay(String dataCriteriaDisplay) {
+    this.dataCriteriaDisplay = dataCriteriaDisplay;
+  }
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/qdm/humanreadable/model/HumanReadableCodeModel.java]: I missed this one. It will also be generated by lombok.

Comment (sb-cecilialiu) [reply src/main/java/gov/cms/madie/qdm/humanreadable/model/HumanReadableCodeModel.java]: yeah just removed it thanks Joe!

File: src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java (hunk 7 above)
Comment (jkotanchik-SB) [reply src/main/java/gov/cms/mat/cql_elm_translation/service/HumanReadableService.java]: I check in the MAT, it is the QDM Data Element portion of the Data Criteria. 

Given `["Encounter, Performed":"Some Direct Reference Code"]`, then "Encounter, Performed" would be the datatype.

File: src/main/java/gov/cms/mat/cql_elm_translation/utils/cql/CQLTools.java (hunk 9 above)
Comment (jkotanchik-SB) [reply src/main/java/gov/cms/mat/cql_elm_translation/utils/cql/CQLTools.java:237]: Double checked, it does not impact Call Hierarchy determination. So we're good. 👍

File: src/test/java/gov/cms/mat/cql_elm_translation/service/HumanReadableServiceTest.java [hunk 11]
```diff
@@ -88,21 +131,124 @@ void setUp() {
...
+        CQLDefinition.builder()
+            .id("MATGlobalCommonFunctionsQDM-1.0.000|Global|NormalizeInterval")
+            .definitionName("NormalizeInterval")
+            .definitionLogic(
+                "define function \"NormalizeInterval\"(pointInTime DateTime, period Interval<DateTime> ):\n  if pointInTime is not null then Interval[pointInTime, pointInTime]\n    else if period is not null then period \n    else null as Interval<DateTime>")
+            .parentLibrary("MATGlobalCommonFunctionsQDM")
+            .isFunction(true)
+            .build();
+    allDefinitions = new HashSet<>(Arrays.asList(definition1, definition2, function, definition3));
+    //    usedFunctions = new HashSet<>(Arrays.asList(function));
```
Comment (jkotanchik-SB) [inline src/test/java/gov/cms/mat/cql_elm_translation/service/HumanReadableServiceTest.java]: Dead code. Remove pls.

**Review by jkotanchik-SB — DISMISSED** [review]

**Review by jkotanchik-SB — APPROVED** [review]
